// @ngx-gallery/lightbox/bundles/ngx-gallery-lightbox.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/animations'), require('@ngx-gallery/core'), require('@angular/cdk/overlay'), require('@angular/cdk/portal'), require('@angular/cdk/keycodes'), require('@angular/common')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/animations', '@ngx-gallery/core', '@angular/cdk/overlay', '@angular/cdk/portal', '@angular/cdk/keycodes', '@angular/common'], factory) :
	(factory((global['ngx-gallery'] = global['ngx-gallery'] || {}, global['ngx-gallery'].lightbox = {}),global.ng.core,global.ng.animations,global.core$1,global.ng.cdk.overlay,global.ng.cdk.portal,global.ng.cdk.keycodes,global.ng.common));
}(this, (function (exports,core,animations,core$1,overlay,portal,keycodes,common) { 'use strict';

var LightboxComponent = /** @class */ (function () {
    function LightboxComponent() {
    }
    LightboxComponent.prototype.ngOnDestroy = function () {
        this.close();
    };
    return LightboxComponent;
}());
LightboxComponent.decorators = [
    { type: core.Component, args: [{
                selector: 'lightbox',
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                preserveWhitespaces: false,
                animations: [
                    animations.trigger('slideGalleryIn', [
                        animations.state('void', animations.style({ transform: 'translateY(25%) scale(0.9)', opacity: 0 })),
                        animations.state('enter', animations.style({ transform: 'translateY(0%) scale(1)', opacity: 1 })),
                        animations.state('exit', animations.style({ transform: 'translateY(25%)', opacity: 0 })),
                        animations.transition('* => *', animations.animate('400ms cubic-bezier(0.25, 0.8, 0.25, 1)')),
                    ])
                ],
                template: "\n    <gallery [@slideGalleryIn] [id]=\"id\" [destroyRef]=\"false\">\n      <div class=\"g-btn-close\" (tapClick)=\"close()\"></div>\n    </gallery>\n  "
            },] },
];
LightboxComponent.ctorParameters = function () { return []; };
var LIGHTBOX_CONFIG = new core.InjectionToken('lightboxConfig');
var defaultConfig = {
    backdropClass: 'g-backdrop',
    panelClass: 'g-overlay',
    hasBackdrop: true,
    keyboardShortcuts: true
};
var Lightbox = /** @class */ (function () {
    function Lightbox(config, _gallery, _overlay) {
        this._gallery = _gallery;
        this._overlay = _overlay;
        this.config = Object.assign({}, defaultConfig, config);
    }
    Lightbox.prototype.setConfig = function (config) {
        this.config = Object.assign({}, this.config, config);
    };
    Lightbox.prototype.open = function (i, id, config) {
        var _this = this;
        if (i === void 0) { i = 0; }
        if (id === void 0) { id = 'lightbox'; }
        var _config = config ? Object.assign({}, this.config, config) : this.config;
        var overlayConfig = {
            backdropClass: _config.backdropClass,
            panelClass: _config.panelClass,
            hasBackdrop: _config.hasBackdrop,
            positionStrategy: this._overlay.position().global().centerHorizontally().centerVertically(),
            scrollStrategy: this._overlay.scrollStrategies.block()
        };
        var galleryRef = this._gallery.ref(id);
        galleryRef.set(i);
        this._overlayRef = this._overlay.create(overlayConfig);
        var galleryPortal = new portal.ComponentPortal(LightboxComponent);
        var compRef = this._overlayRef.attach(galleryPortal);
        compRef.instance.id = id;
        compRef.instance.close = function () { return _this.close(); };
        if (_config.hasBackdrop) {
            this._overlayRef.backdropClick().subscribe(function () { return _this.close(); });
        }
        if (_config.keyboardShortcuts) {
            this._overlayRef.keydownEvents().subscribe(function (event) {
                switch (event.keyCode) {
                    case keycodes.LEFT_ARROW:
                        galleryRef.prev(id);
                        break;
                    case keycodes.RIGHT_ARROW:
                        galleryRef.next(id);
                        break;
                    case keycodes.ESCAPE:
                        _this.close();
                }
            });
        }
    };
    Lightbox.prototype.close = function () {
        if (this._overlayRef.hasAttached()) {
            this._overlayRef.dispose();
        }
    };
    return Lightbox;
}());
Lightbox.decorators = [
    { type: core.Injectable },
];
Lightbox.ctorParameters = function () { return [
    { type: undefined, decorators: [{ type: core.Inject, args: [LIGHTBOX_CONFIG,] },] },
    { type: core$1.Gallery, },
    { type: overlay.Overlay, },
]; };
function lightboxFactory(config, gallery, overlay$$1) {
    return new Lightbox(config, gallery, overlay$$1);
}
var LightboxModule = /** @class */ (function () {
    function LightboxModule() {
    }
    LightboxModule.forRoot = function (config) {
        return {
            ngModule: LightboxModule,
            providers: [
                {
                    provide: LIGHTBOX_CONFIG,
                    useValue: config
                },
                {
                    provide: Lightbox,
                    useFactory: lightboxFactory,
                    deps: [LIGHTBOX_CONFIG, core$1.Gallery, overlay.Overlay]
                }
            ]
        };
    };
    return LightboxModule;
}());
LightboxModule.decorators = [
    { type: core.NgModule, args: [{
                imports: [
                    common.CommonModule,
                    overlay.OverlayModule,
                    core$1.GalleryModule
                ],
                declarations: [
                    LightboxComponent
                ],
                entryComponents: [
                    LightboxComponent
                ]
            },] },
];
LightboxModule.ctorParameters = function () { return []; };

exports.LightboxComponent = LightboxComponent;
exports.Lightbox = Lightbox;
exports.lightboxFactory = lightboxFactory;
exports.LightboxModule = LightboxModule;
exports.ɵa = LIGHTBOX_CONFIG;

Object.defineProperty(exports, '__esModule', { value: true });

})));
